pallets/asset-registry/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+	pub enum Event<T: Config> {
+		ReserveAssetRegistered { asset_id: AssetIdOf<T>, asset: MultiLocation },
+		ReserveAssetUnregistered { asset_id: AssetIdOf<T>, asset_multi_location: MultiLocation },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The Asset ID is already registered
+		AssetAlreadyRegistered,
+		/// The Asset ID does not exist
+		AssetDoesNotExist,
+		/// The Asset ID is not Registered
+		AssetIsNotRegistered,
+		/// Invalid MultiLocation
+		WrongMultiLocation,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::register_reserve_asset())]
+		pub fn register_reserve_asset(
+			origin: OriginFor<T>,
+			asset_id: AssetIdOf<T>,
+			asset_multi_location: MultiLocation,
+		) -> DispatchResult {
+			T::ReserveAssetModifierOrigin::ensure_origin(origin)?;
+
+			// verify asset exists on pallet-assets
+			ensure!(Self::asset_exists(asset_id), Error::<T>::AssetDoesNotExist);
+
+			// verify asset is not yet registered
+			ensure!(
+				AssetIdMultiLocation::<T>::get(&asset_id).is_none(),
+				Error::<T>::AssetAlreadyRegistered
+			);
+
+			// verify MultiLocation is valid
+			let parents_multi_location_ok = { asset_multi_location.parents == 1 };
+			let junctions_multi_location_ok = matches!(asset_multi_location.interior, Junctions::X3(Parachain(_), PalletInstance(_), GeneralIndex(_)));
+
+			ensure!(
+				parents_multi_location_ok && junctions_multi_location_ok,
+				Error::<T>::WrongMultiLocation
+			);
+
+			// register asset
+			AssetIdMultiLocation::<T>::insert(&asset_id, &asset_multi_location);
+			AssetMultiLocationId::<T>::insert(&asset_multi_location, &asset_id);
+
+			Self::deposit_event(Event::ReserveAssetRegistered {
+				asset_id,
+				asset: asset_multi_location,
+			});
+
+			Ok(())
+		}
+
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::unregister_reserve_asset())]
+		pub fn unregister_reserve_asset(
+			origin: OriginFor<T>,
+			asset_id: AssetIdOf<T>,
+		) -> DispatchResult {
+			T::ReserveAssetModifierOrigin::ensure_origin(origin)?;
+
+			// verify asset is registered
+			let asset_multi_location = AssetIdMultiLocation::<T>::get(&asset_id)
+				.ok_or(Error::<T>::AssetIsNotRegistered)?;
+
+			// unregister asset
+			AssetIdMultiLocation::<T>::remove(&asset_id);
+			AssetMultiLocationId::<T>::remove(&asset_multi_location);
+
+			Self::deposit_event(Event::ReserveAssetUnregistered { asset_id, asset_multi_location });
+			Ok(())
+		}
+	}
+
+	impl<T: Config> xcm_primitives::AssetMultiLocationGetter<AssetIdOf<T>> for Pallet<T> {
+		fn get_asset_multi_location(asset_id: AssetIdOf<T>) -> Option<MultiLocation> {
+			AssetIdMultiLocation::<T>::get(asset_id)
+		}
+
+		fn get_asset_id(asset_type: MultiLocation) -> Option<AssetIdOf<T>> {
+			AssetMultiLocationId::<T>::get(asset_type)
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn asset_exists(asset_id: AssetIdOf<T>) -> bool {
+			!T::Assets::minimum_balance(asset_id).is_zero()
```

Comment:
This is a bit of a trick implementation. I think a short comment explaining why it's implemented like this would be helpful for a reader who doesn't know pallet-assets well.